spec/src/main/asciidoc/variable-expansion.asciidoc

```diff
@@ -0,0 +1,137 @@
+`META-INF/microprofile-config.properties`). Its value will be used by the MicroProfile Config implementation to _globally_ enable or disable variable expansion.
+In the absence of this property, variable expansion is enabled.
+====
+
+---
+
+
+=== Variable Syntax
+
+The syntax to use a variable is `${key}` for a _required_ config property and `${<key>:<default_value>}` for
+an _optional_ config property.
+
+If the variable is required , by using `${key}`, a config property with the name `key` MUST be present in the Config to be expanded, otherwise, the
+Config API will throw a `NoSuchElementException` (or an `Optional.empty()` value).
+
+For example, let's define the configuration properties:
+
+[source]
+----
+server.url=http://${server.host}:{server.port}/endpoint
+server.host=example.org
+----
+
+A call to `Config.getValue("server.url", String.class)` will throw a `NoSuchElementException` since the
+`server.port` property is not configured.
+Likewise, a call to `Config.getOptionalValue("server.url", String.class)` will return an `Optional.empty()`.
+
+==== Default value
+
+A variable can define a _default_ value to be expanded if the corresponding config property is not present.
+
+For example, let's define the configuration properties:
+
+[source]
+----
+server.url=http://${server.host:example.org}:{server.port:8080}/endpoint
+----
+
+If neither `server.host` nor `server.port` config properties are present, the Config API will use the variable's default values and will return
+`http://example.org:8080/endpoint` for the `server.url`.
+
+It is possible to define an empty default value for a variable.
+If the key is not present in the Config, it will not be expanded.
+
+For example, let's define the configuration properties:
+
+[source]
+----
+server.url=http://example.org:8080/endpoint#${server.anchor:}
+----
+
+When the `server.anchor` configuration property is not present, the Config API will return
+`http://example.org:8080/endpoint#` for the `server.url`.
+
+If the `server.anchor` is configured with the value `id123`, the Config API will return
+`http://example.org:8080/endpoint#id123` for the `server.url`.
+
+A default value does not support variable expansion (a default value is _raw_).
+
+For example, let's define the configuration properties:
+
+[source]
+----
+foo.one=Hello
+foo.two=${foo.three:${foo.one}}
+----
+
+`foo.two` defines a variable which key is `foo.three` and which default value is `${foo.one}`.
+If the config property `foo.three` is not present, the Config API will return `${foo.one}` for the `foo.two` config property.
+
+=== Recursive variable replacement
+
+It is possible to have two configuration properties referencing each other with variables:
+
+[source]
+----
+prop1 = ${prop2}
+prop2 = ${prop1}
+----
+
+Such infinite variable expansion is not supported and the Config API will throw a `NoSuchElementException` (or
+return an `Optional.empty()`) value for both properties.
```

Comment:
This requires a possibly expensive "visited" set to be maintained during expansion.  Could there just be a simple numerical recursion limit instead?

Reply:
Also returning `empty` is potentially hazardous: I think it would be better to flag the error rather than treat the property as missing.  Treating it as missing might result in subtle problems later, whereas an expansion failure resulting in a different exception (say, `IllegalStateException`) would indicate the problem sooner (fail-fast).

Reply:
@dmlloyd I will rephrase it. "Infinite" variable expansion is not correct, something along the lines of "if variable expansion can not be resolved in a limited steps, etc."

You're right about failing fast, raising an exception should be the same for optional or required values. I'll update the PR

Reply:
OK.  What about the idea of using a numerical recursion limit, to avoid having to maintain a visited set?

Reply:
I think the mechanism to detect such recursion (visited set or recursion limit) should not be mandated by the spec and be up to the implementation.
TCK should just check that the implementation raises an exception for such use case.

I'll remove this section from the spec and just keep the TCK test.